pimd/pim_cmd_common.c

```diff
@@ -653,3 +673,1302 @@ int pim_process_no_rp_plist_cmd(struct vty *vty, const char *rp_str,
+	json_object_string_addf(json_row, "source", "%pPAs", &ch->sg.src);
+	json_object_string_addf(json_row, "group", "%pPAs", &ch->sg.grp);
+	json_object_string_add(json_row, "localMembership",
+			       ch->local_ifmembership == PIM_IFMEMBERSHIP_NOINFO
+					   ? "NOINFO"
+					   : "INCLUDE");
+	json_object_object_add(json_iface, ch_grp_str, json_row);
+}
+
+void pim_show_membership(struct pim_instance *pim, struct vty *vty, bool uj)
+{
+	struct pim_interface *pim_ifp;
+	struct pim_ifchannel *ch;
+	struct interface *ifp;
+	enum json_type type;
+	json_object *json = NULL;
+	json_object *json_tmp = NULL;
+
+	json = json_object_new_object();
+
+	FOR_ALL_INTERFACES (pim->vrf, ifp) {
+		pim_ifp = ifp->info;
+		if (!pim_ifp)
+			continue;
+
+		RB_FOREACH (ch, pim_ifchannel_rb, &pim_ifp->ifchannel_rb) {
+			pim_show_membership_helper(vty, pim_ifp, ch, json);
+		} /* scan interface channels */
+	}
+
+	if (uj) {
+		vty_json(vty, json);
+	} else {
+		vty_out(vty,
+			"Interface         Address          Source           Group            Membership\n");
+
+		/*
+		 * Example of the json data we are traversing
+		 *
+		 * {
+		 *   "swp3":{
+		 *     "name":"swp3",
+		 *     "state":"up",
+		 *     "address":"10.1.20.1",
+		 *     "index":5,
+		 *     "flagMulticast":true,
+		 *     "flagBroadcast":true,
+		 *     "lanDelayEnabled":true,
+		 *     "226.10.10.10":{
+		 *       "source":"*",
+		 *       "group":"226.10.10.10",
+		 *       "localMembership":"INCLUDE"
+		 *     }
+		 *   }
+		 * }
+		 */
+
+		/* foreach interface */
+		json_object_object_foreach(json, key, val)
+		{
+
+			/* Find all of the keys where the val is an object. In
+			 * the example
+			 * above the only one is 226.10.10.10
+			 */
+			json_object_object_foreach(val, if_field_key,
+						   if_field_val)
+			{
+				type = json_object_get_type(if_field_val);
+
+				if (type == json_type_object) {
+					vty_out(vty, "%-16s  ", key);
+
+					json_object_object_get_ex(
+						val, "address", &json_tmp);
+					vty_out(vty, "%-15s  ",
+						json_object_get_string(
+							json_tmp));
+
+					json_object_object_get_ex(if_field_val,
+								  "source",
+								  &json_tmp);
+					vty_out(vty, "%-15s  ",
+						json_object_get_string(
+							json_tmp));
+
+					/* Group */
+					vty_out(vty, "%-15s  ", if_field_key);
+
+					json_object_object_get_ex(
+						if_field_val, "localMembership",
+						&json_tmp);
+					vty_out(vty, "%-10s\n",
+						json_object_get_string(
+							json_tmp));
+				}
+			}
+		}
+		json_object_free(json);
+	}
+}
+
+static void pim_show_channel_helper(struct pim_instance *pim, struct vty *vty,
+				    struct pim_interface *pim_ifp,
+				    struct pim_ifchannel *ch, json_object *json,
+				    bool uj)
+{
+	struct pim_upstream *up = ch->upstream;
+	json_object *json_group = NULL;
+	json_object *json_row = NULL;
+
+	if (uj) {
+		char grp_str[PIM_ADDRSTRLEN];
+		char src_str[PIM_ADDRSTRLEN];
+
+		snprintfrr(grp_str, sizeof(grp_str), "%pPAs", &up->sg.grp);
```

Comment:
format directly in `json_object_*` calls

Reply:
there is no f function for json_object_object_get_ex, hence using snprintfrr here